### DISCUSSION TITLE
How to boost::serialize S2PointIndex

### DISCUSSION THREAD

**Discussion by GoroYeh-HRI** [google/s2geometry#342] [Q&A]
Hi,

If I want to use `boost::serialize()` to store class `S2PointIndex` as binary stream,
how do I do it?
Do I need to implement my own serialization function for `S2PointIndex` ?
Thanks.

**Comment by smcallis** [comment]
It looks like we don't have an Encode() method on that one, so I think the
answer is yes.  Fortunately it's pretty simple under the hood.

**Comment by GoroYeh-HRI** [comment]
Hi @smcallis , thanks for your reply.
Could you elaborate more on the way to do it?

I'm currently tracing class `S2PointIndex` and think I need to serialize `S2Point`, `S2CellId`, and `s2internal::BTreeMultimap`.
The last one `BTreeMultimap` is a little bit complicated, so I'm stuck at this point.

Could you provide more  details on how to implement serialize() function for this class?
Thank you very much!

**Comment by smcallis** [comment]
S2Point and S2CellId already have Encode/Init methods so they can serialize themselves for you.  s2internal::BTreeMultimap is just absl::btree_multimap.  I'd probably just iterate that container and store pairs of (S2CellId, [list of data]), then read that back into a new absl::btree_multimap when decoding.  That might not be the absolute most efficient encoding space-wise but it is simple.  I'd worry more about the fact that PointIndex takes a template parameter for the data, so you'll have to serialize the data type too.

**Comment by GoroYeh-HRI** [comment]
> S2Point and S2CellId already have Encode/Init methods so they can serialize themselves for you. s2internal::BTreeMultimap is just absl::btree_multimap. I'd probably just iterate that container and store pairs of (S2CellId, [list of data]), then read that back into a new absl::btree_multimap when decoding. That might not be the absolute most efficient encoding space-wise but it is simple. I'd worry more about the fact that PointIndex takes a template parameter for the data, so you'll have to serialize the data type too.

Thank you so much for your reply!
So your point is, for S2Point and S2CellId I don't need to write any codes to serialize them?
And for s2internal::BTreeMultimap, in addition to the container I need to handle the parameter.

I noticed the BTreeMultimap is instantiated like this:
`container_internal::btree<container_internal::
map_params<Key, Value, Compare, Alloc,/*TargetNodeSize=*/256,/*IsMulti=*/true>>`

Did you mean I'll need to serialize `map_params<>` ? 
I noticed map_params inherits common_params which inherits common_policy_traits.
I wonder how deep I should trace in order to successfully store all the necessary data.
Thank you!

**Comment by smcallis** [comment]
You'll have to write code to hook S2Point and S2CellId into boost::serialize, but you can probably do that using their existing Encode/Init methods.  I don't think you'll need to worry about storing the map params, that's a runtime detail different from storing the data in a serialized form.
